### PR TITLE
Refactor and optimize `SwapMath`

### PR DESCRIPTION
**Pull request by shuhuiluo**
## Related Issue
Which issue does this pull request resolve?

Obscure control flow related to `SwapMath`.

## Description of changes

Improvements to the `SwapMath` library have been implemented. A new function, `getSqrtRatioTarget`, written in `Yul` using bitwise operations, has been introduced to replace a nested ternary operation within `Pool.swap`:
```Solidity
(
    params.zeroForOne
        ? step.sqrtPriceNextX96 < params.sqrtPriceLimitX96
        : step.sqrtPriceNextX96 > params.sqrtPriceLimitX96
) ? params.sqrtPriceLimitX96 : step.sqrtPriceNextX96
```
which basically was doing
```Solidity
zeroForOne ? max(sqrtPriceNextX96, sqrtPriceLimitX96) : min(sqrtPriceNextX96, sqrtPriceLimitX96)
```

Furthermore, the `computeSwapStep` function has been refactored, resulting in a cleaner control flow and significant gas savings. The gas savings achieved range from 10% to 28%, as can be seen in `SwapMath.spec.ts.snap`. In conjunction with the changes introduced in pull request #258, the maximum savings on `computeSwapStep` can reach up to 40%.

All tests pass except for `SwapMath::#computeSwapStep::entire input amount taken as fee`, where `amountIn` changes from `0` to `9`, and `feeAmount` changes from `10` to `1`.

I wonder if "entire input amount taken as fee when output is zero" is an intended feature and the invariant `amountIn == 0 && feeAmount == 10` should be kept. Adjusting the invariant to `amountIn + feeAmount == 10` appears to be acceptable and reasonable.

### CODE REVIEW

**Comment by hensha256** [comment]
Please can you merge main so we can re-review the gas improvements 🙏

**Comment by shuhuiluo** [comment]
> Please can you merge main so we can re-review the gas improvements 🙏

@hensha256 Done. As stated before, I had to modify `test_entireInputAmountTakenAsFee`, which I believe to be an artifact of the previous control flow instead of an intended feature.

File: src/libraries/SwapMath.sol [hunk 1]
```diff
@@ -49,63 +51,54 @@ library SwapMath {
...
                     ? SqrtPriceMath.getAmount1Delta(sqrtPriceTargetX96, sqrtPriceCurrentX96, liquidity, false)
                     : SqrtPriceMath.getAmount0Delta(sqrtPriceCurrentX96, sqrtPriceTargetX96, liquidity, false);
                 if (uint256(amountRemaining) >= amountOut) {
+                    // `amountOut` is capped by the target price
                     sqrtPriceNextX96 = sqrtPriceTargetX96;
                 } else {
+                    // cap the output amount to not exceed the remaining output amount
+                    amountOut = uint256(amountRemaining);
                     sqrtPriceNextX96 = SqrtPriceMath.getNextSqrtPriceFromOutput(
                         sqrtPriceCurrentX96, liquidity, uint256(amountRemaining), zeroForOne
```
Comment (hensha256) [inline src/libraries/SwapMath.sol:94]: is it cheaper to pass in `amountOut` to prevent double-casting `amountRemaining`?

Comment (shuhuiluo) [reply src/libraries/SwapMath.sol:94]: Casting between `uint256` and `int256` shouldn't matter.

Comment (snreynolds) [reply src/libraries/SwapMath.sol:94]: was gonna ask same thing - even if it doesn't make a difference I think the code reads better with using the variable we just defined above imo

File: src/libraries/SwapMath.sol [hunk 2]
```diff
@@ -49,63 +51,54 @@ library SwapMath {
...
+                    FullMath.mulDiv(uint256(-amountRemaining), MAX_FEE_PIPS - _feePips, MAX_FEE_PIPS);
                 amountIn = zeroForOne
                     ? SqrtPriceMath.getAmount0Delta(sqrtPriceTargetX96, sqrtPriceCurrentX96, liquidity, true)
                     : SqrtPriceMath.getAmount1Delta(sqrtPriceCurrentX96, sqrtPriceTargetX96, liquidity, true);
                 if (amountRemainingLessFee >= amountIn) {
+                    // `amountIn` is capped by the target price
                     sqrtPriceNextX96 = sqrtPriceTargetX96;
+                    feeAmount = _feePips == MAX_FEE_PIPS
+                        ? amountIn
+                        : FullMath.mulDivRoundingUp(amountIn, _feePips, MAX_FEE_PIPS - _feePips);
```
[the remark this answers is not in the document]
Comment (shuhuiluo) [reply src/libraries/SwapMath.sol:69]: Caching `MAX_FEE_PIPS - _feePips` increases runtime gas while saving bytecode size.

File: src/libraries/SwapMath.sol (hunk 1 above)
Comment (shuhuiluo) [reply src/libraries/SwapMath.sol]: Passing `amountOut` to `getNextSqrtPriceFromOutput`.

**Review by hensha256 — APPROVED** [review]

**Pull request merged**